lib/ash/type/map.ex

```diff
@@ -16,26 +54,69 @@ defmodule Ash.Type.Map do
 
   def cast_input(nil, _), do: {:ok, nil}
 
-  def cast_input(value, _) when is_binary(value) do
+  def cast_input(value, constraints) when is_binary(value) do
     case Jason.decode(value) do
       {:ok, value} ->
-        {:ok, value}
+        check_constraints(value, constraints)
 
       _ ->
         :error
     end
   end
 
-  def cast_input(value, _) when is_map(value), do: {:ok, value}
+  def cast_input(value, constraints) when is_map(value), do: check_constraints(value, constraints)
   def cast_input(_, _), do: :error
 
   @impl true
   def cast_stored(nil, _), do: {:ok, nil}
-  def cast_stored(value, _) when is_map(value), do: {:ok, value}
+
+  def cast_stored(value, constraints) when is_map(value),
+    do: check_constraints(value, constraints)
```

Comment:
I don't know if this is needed here. Theoretically, a stored value should already be validated, but if other services use the same DB, for example, it could be that it's not correct.

Reply:
Yeah, we should leave it out of here because not all types can necessarily expect to be `cast_input` from the result of `cast_stored` (like it might be a one way translation). We could potentially add that as a constraint/option later down the road if people need it.